src/Utilities/cachingoptimizer.jl

```diff
@@ -227,54 +227,75 @@ function MOI.addconstraint!(m::CachingOptimizer, func::MOI.AbstractFunction, set
     return cindex
 end
 
-function MOI.canmodifyconstraint(m::CachingOptimizer, cindex::CI, change)
-    MOI.canmodifyconstraint(m.model_cache, cindex, change) || return false
+function MOI.canmodify(m::CachingOptimizer, ::Type{C}, change) where C <: CI
+    MOI.canmodify(m.model_cache, C, change) || return false
     if m.state == AttachedOptimizer && m.mode == Manual
-        MOI.canmodifyconstraint(m.optimizer, m.model_to_optimizer_map[cindex], change) || return false
+        MOI.canmodify(m.optimizer, C, change) || return false
     end
     return true
 end
 
-function MOI.modifyconstraint!(m::CachingOptimizer, cindex::CI, change)
-    if m.mode == Automatic && m.state == AttachedOptimizer && !MOI.canmodifyconstraint(m.optimizer, cindex, typeof(change))
+function MOI.modify!(m::CachingOptimizer, cindex::CI, change)
+    if m.mode == Automatic && m.state == AttachedOptimizer && !MOI.canmodify(m.optimizer, typeof(cindex), typeof(change))
         resetoptimizer!(m)
     end
-    @assert MOI.canmodifyconstraint(m, cindex, typeof(change))
-    MOI.modifyconstraint!(m.model_cache, cindex, change)
+    @assert MOI.canmodify(m, typeof(cindex), typeof(change))
+    MOI.modify!(m.model_cache, cindex, change)
     if m.state == AttachedOptimizer
-        MOI.modifyconstraint!(m.optimizer, m.model_to_optimizer_map[cindex], mapvariables(m.model_to_optimizer_map,change))
+        MOI.modify!(m.optimizer, m.model_to_optimizer_map[cindex], mapvariables(m.model_to_optimizer_map,change))
     end
     return
 end
 
-function MOI.modifyconstraint!(m::CachingOptimizer, cindex::CI{F,S}, set::S) where {F<:MOI.AbstractFunction, S<:MOI.AbstractSet}
-    if m.mode == Automatic && m.state == AttachedOptimizer && !MOI.canmodifyconstraint(m.optimizer, cindex, S)
+function MOI.canset(m::CachingOptimizer, attr::Union{MOI.ConstraintFunction, MOI.ConstraintSet}, ::Type{C}) where C <: CI
+    if !MOI.canset(m.model_cache, attr, C)
+        return false
+    end
+    if m.state == AttachedOptimizer && m.mode == Manual
+        if !MOI.canset(m.optimizer, attr, C)
+            return false
+        end
+    end
+    return true
+end
+
+# This function avoids duplicating code, but allows us to strongly type the
+# arguments of the set! functions below.
```

Comment:
I didn't get that immediately, mentioning the 4th argument explicitly in the comment would be clearer